src/Redirection/Dynamic/RedirectionManager.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace AppBundle\Redirection\Dynamic;
+
+use AppBundle\Entity\Redirection;
+use Doctrine\ORM\EntityManagerInterface;
+use JMS\Serializer\Serializer;
+use Psr\Cache\CacheItemPoolInterface;
+
+class RedirectionManager
+{
+    private $cache;
+    private $entityManager;
+    private $serializer;
+
+    public function __construct(CacheItemPoolInterface $cache, EntityManagerInterface $entityManager, Serializer $serializer)
+    {
+        $this->cache = $cache;
+        $this->entityManager = $entityManager;
+        $this->serializer = $serializer;
+    }
+
+    public function refreshRedirectionCache(Redirection $redirection): void
+    {
+        $item = $this->cache
+            ->getItem(md5($redirection->getFrom()))
+            ->set($this->serializer->serialize($redirection, 'json'))
+        ;
+        $this->cache->save($item);
+    }
+
+    public function setRedirection(string $source, string $target, int $type = 301): void
+    {
+        if (!$redirection = $this->entityManager->getRepository(Redirection::class)->findOneByOriginUri($source)) {
+            $redirection = new Redirection();
+            $this->entityManager->persist($redirection);
+        }
+        $redirection->setFrom($source);
+        $redirection->setTo($target);
+        $redirection->setType($type);
+
+        $this->entityManager->flush();
+    }
+
+    public function removeRedirection(string $source): void
+    {
+        if ($redirection = $this->entityManager->getRepository(Redirection::class)->findOneByOriginUri($source)) {
+            $this->entityManager->remove($redirection);
+            $this->entityManager->flush();
+            $this->cache->deleteItem(md5($source));
+        }
+    }
+
+    public function getRedirection(string $source): ?Redirection
+    {
+        if (($item = $this->cache->getItem(md5($source))) && $item->isHit()) {
```

Comment:
`&& $item->isHit()` : here `$item` is not assigned yet and therefore is undefined.

This should be:
```
$item = $this->cache->getItem(md5($source));

if ($item && $item->isHit()) {
```

Reply:
my bad. i did not see the extra brackets 🤕

Reply:
It's not very readable though. I prefer the way @Remg suggested.